File: leetcode/leecode207/solution.ts
/**
 * 考点：DFS、BFS、图、拓扑排序
 * 题目：Course Schedule（课程表）
 * 题目描述：有 numCourses 门课，prerequisites[i]=[a,b] 表示上 a 前必须先上 b。判断能否完成所有课程（判环）。
 * 示例：numCourses=2,prerequisites=[[1,0],[0,1]] 输出 false
 * 思路：BFS Kahn 算法——构建邻接表和入度数组，入度为 0 的入队，BFS 消去入度，最终比较出队数与课程数。
 * 时间复杂度：O(V + E)
 * 空间复杂度：O(V + E)
 */
function canFinish(numCourses: number, prerequisites: number[][]): boolean {
    const adj: number[][] = Array.from({ length: numCourses }, () => []);
    const indegree: number[] = new Array(numCourses).fill(0);

    for (const [a, b] of prerequisites) {
        adj[b].push(a);
        indegree[a]++;
    }

    const queue: number[] = [];
    for (let i = 0; i < numCourses; i++) {
        if (indegree[i] === 0) queue.push(i);
    }

    let count = 0;
    while (queue.length > 0) {
        const cur = queue.shift()!;
        count++;
        for (const next of adj[cur]) {
            indegree[next]--;
            if (indegree[next] === 0) queue.push(next);
        }
    }

    return count === numCourses;
}
export { canFinish };

File: leetcode/leecode207/solution.py
"""
考点：DFS、BFS、图、拓扑排序
题目：Course Schedule（课程表）
思路：BFS Kahn 算法——构建邻接表和入度数组，入度为 0 的入队，BFS 消去入度，
      最终比较出队数与课程数。
时间复杂度：O(V + E)
空间复杂度：O(V + E)
"""

from collections import deque


def canFinish(numCourses: int, prerequisites: list[list[int]]) -> bool:
    # 构建邻接表：adj[b] 存所有依赖 b 的课程（即 a 依赖 b，b→a）
    adj: list[list[int]] = [[] for _ in range(numCourses)]
    indegree = [0] * numCourses  # 入度数组

    for a, b in prerequisites:
        adj[b].append(a)  # b → a
        indegree[a] += 1  # a 的入度 +1

    # 所有入度为 0 的课程入队（不需要先修课程）
    q = deque(i for i in range(numCourses) if indegree[i] == 0)

    count = 0  # 已完成的课程数
    while q:
        cur = q.popleft()  # Python deque 的 popleft 是 O(1)
        count += 1
        # 遍历当前课程的所有后续课程
        for nxt in adj[cur]:
            indegree[nxt] -= 1
            if indegree[nxt] == 0:
                q.append(nxt)

    return count == numCourses  # 所有课程都能完成则无环


if __name__ == "__main__":
    # 示例 1: 2, [[1,0]] → true (0→1 可行)
    assert canFinish(2, [[1, 0]]) is True
    # 示例 2: 2, [[1,0],[0,1]] → false (循环依赖)
    assert canFinish(2, [[1, 0], [0, 1]]) is False
    print("全部测试通过")
